rep-0199.rst

```diff
@@ -0,0 +1,407 @@
+REP: 199
+Title: Coordinate Frames for Serial Industrial Manipulators
+Author: G.A. vd. Hoorn <g.a.vanderhoorn@tudelft.nl>
+Status: Draft
+Type: Informational
+Content-Type: text/x-rst
+Requires: 103
+Created: 11-Jan-2019
+
+
+Abstract
+========
+
+This REP specifies naming conventions and semantics for coordinate frames and links of serial industrial manipulators that are used with ROS.
+
+
+Motivation
+==========
+
+Developers of packages aimed at serial industrial manipulators need a shared convention for coordinate frames in order to better integrate and re-use software components.
+Consistent naming of frames and standardisation of semantics allows for re-use of not only software but also of experience and tooling.
+
+In addition, commercially available industrial robot controllers already define a number of Cartesian coordinate frames for users to define poses in and to refer to when manually controlling the robot (i.e.: while jogging or programming using teach-in).
+In order to make it possible for ROS applications to refer to these coordinate frames in a controller and vendor-agnostic way, this REP proposes a set of generalised frame names with associated semantics that allow for a correspondence between controller internal frames and the (TF) frames that a ROS application communicating with it might use.
+
+REP 105 [#REP105]_ and REP 120 [#REP120]_ already define frames for mobile bases and humanoid robots.
+This REP extends the set of standardised frames in ROS and labels important locations on serial industrial manipulators such as the origin of the main Cartesian coordinate system, the flange and the tool centre point.
+Finally, it provides naming guidelines for intermediate link frames and application-specific tool frames.
+
+
+Definitions
+===========
+
+End Effector (EEF)
+    A device at the end of a robotic arm, designed to interact with the environment.
+    The exact nature of this device depends on the application of the robot.
+Tool Centre Point (TCP)
+    Point in relation to which all robot positioning is defined or the origin of the tool's coordinate frame.
+
+
+Assumptions
+===========
+
+#. The key words *must*, *must not*, *required*, *shall*, *shall not*, *should*, *should not*, *recommended*, *may*, and *optional* in this document are to be interpreted as defined in [#RFC2119]_.
+
+
+Specification
+=============
+
+Coordinate Frames
+-----------------
+
+base_link
+'''''''''
+
+The coordinate frame called ``base_link`` shall be rigidly attached to the robot root body.
```

Comment:
Which frames are optional and which ones mandatory? What happens, if a robot is missing some of the frames? Are there default assumtions such as `base=base_link`, `tool0=flange`? This would allow easier transitions of existing robot models.

```suggestion
The robot model shall define a  coordinate frame called ``base_link``.
```

Reply:
> Which frames are optional and which ones mandatory?

I'll update the REP, but only `base`, `flange` and `tool0` are mandatory. The `link_n` frame names are a result of an earlier review of the REP in which we discussed whether introducing naming conventions for intermediate frames (ie: those of the links) would make sense. In the end I decided that, yes, it does make sense -- for the reasons stated in the rationale paragraph of the `link_n` section.

> What happens, if a robot is missing some of the frames?

similar to what would happen if you intend to reuse the navigation stack with a robot that does not have an `odom` frame: things will fail, as the navstack requires a REP-105 compliant system.

> Are there default assumtions such as `base=base_link`, `tool0=flange`? This would allow easier transitions of existing robot models.

While the idea of a default is nice, I'm curious how you see this working. The REP attempts to standardise urdf/xacro `link` / TF frame names. There is no facility in either of those for defaults or fallbacks.

I would also not be very comfortable if there were: without these frames present, a TF lookup `base -> tool0` could evaluate down to `base_link -> flange` (if `flange` exists), which is an entirely different lookup and the result would be completely different, semantically, from the intended lookup (as `base_link` could be anywhere and has no orientation constraints and `flange` intentionally does not correspond to `tool0`).

Finally: as I wrote above: the only frames that are *really* required are `base` and `tool0` (and potentially `flange`). If we go with the idea of defaults, just about every robot model could start claiming compliance with this REP after adding a `link` with the name `flange` somewhere. That is not the intention of the REP.

---

These are good questions btw, and I'll see how I can incorporate them (and their answers) into the REP proper.